src/Microsoft.Extensions.HealthChecks/HealthCheckService.cs

```diff
@@ -1,20 +1,113 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
 using System.Collections.Generic;
+using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
 
 namespace Microsoft.Extensions.HealthChecks
 {
     public class HealthCheckService : IHealthCheckService
     {
-        public IReadOnlyDictionary<string, IHealthCheck> _checks;
+        private readonly HealthCheckBuilder _builder;
+        private readonly IReadOnlyList<HealthCheckGroup> _groups;
+        private readonly HealthCheckGroup _root;
+        private readonly IServiceProvider _serviceProvider;
+        private readonly IServiceScopeFactory _serviceScopeFactory;
 
-        public HealthCheckService(HealthCheckBuilder builder)
-            => _checks = builder.Checks;
+        public HealthCheckService(HealthCheckBuilder builder, IServiceProvider serviceProvider, IServiceScopeFactory serviceScopeFactory)
+        {
+            _builder = builder;
+            _groups = GetGroups().Where(group => group.GroupName != string.Empty).ToList();
+            _root = GetGroup(string.Empty);
+            _serviceProvider = serviceProvider;
+            _serviceScopeFactory = serviceScopeFactory;
+        }
 
-        public Task<CompositeHealthCheckResult> CheckHealthAsync(CheckStatus partiallyHealthyStatus, CancellationToken cancellationToken)
-            => CheckExecutor.RunChecksAsync(_checks, partiallyHealthyStatus, cancellationToken);
+        public async Task<CompositeHealthCheckResult> CheckHealthAsync(CancellationToken cancellationToken)
+        {
+            using (var scope = GetServiceScope())
+            {
+                var scopeServiceProvider = scope.ServiceProvider;
+                var groupTasks = _groups.Select(group => new { Group = group, Task = RunGroupAsync(scopeServiceProvider, group, cancellationToken) }).ToList();
+                var result = await RunGroupAsync(scopeServiceProvider, _root, cancellationToken).ConfigureAwait(false);
+
+                await Task.WhenAll(groupTasks.Select(x => x.Task));
+
+                foreach (var groupTask in groupTasks)
+                    result.Add($"Group({groupTask.Group.GroupName})", groupTask.Task.Result);
```

Comment:
Please add braces to these. Don't want a `goto fail` error here!

Applies to a few other places too.